### PR TITLE
fixes #22647 - ensure sync notification on repo sync

### PR DESCRIPTION
**Pull request by stbenjam**

### CODE REVIEW

**Comment by stbenjam** [comment]
[test katello]

**Comment by stbenjam** [comment]
[test katello]

**Comment by ehelms** [comment]
[test katello]

**Comment by ehelms** [comment]
@stbenjam Seems it ran OK this time

**Comment by stbenjam** [comment]
[test katello]

**Comment by stbenjam** [comment]
[test katello]

**Comment by stbenjam** [comment]
[test katello]

**Comment by stbenjam** [comment]
[test katello]

**Comment by jlsherrill** [comment]
[test katello]  23rd time's the charm!

**Comment by stbenjam** [comment]
Guess not:

```

wrapper script does not seem to be touching the log file in /var/lib/workspace/workspace/katello-pr-test/foreman@tmp/durable-c43208a0

(JENKINS-48300: if on a laggy filesystem, consider -Dorg.jenkinsci.plugins.durabletask.BourneShellScript.HEARTBEAT_CHECK_INTERVAL=300)

script returned exit code -1
```

**Comment by stbenjam** [comment]
[test katello]

**Comment by stbenjam** [comment]
FINALLY green.  

@jlsherrill Do you think this is the right approach? I wonder if I couldn't do the ensure_sync_notification as an Action that gets planned instead. It might prevent this from causing failures later on for other people writing tests that triggers a sync to be planned? A lot of tests trigger this.

**Comment by jlsherrill** [comment]
@stbenjam i could see argument either way, but i think i would prefer putting it in its own action (although i'm okay leaving it as it is).

**Comment by stbenjam** [comment]
@jlsherrill Moved it to a task.  Does it look right?

File: app/lib/actions/pulp/repository/ensure_sync_notification.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+          end
+
+          # only create a notifier if one doesn't exist with the correct url
+          exists = notifs.select { |n| n['event_types'] == [type] && n['notifier_config']['url'] == url }
+          resource.create(Runcible::Resources::EventNotifier::NotifierTypes::REST_API, {:url => url}, [type]) if exists.empty?
+        end
+      end
+    end
+  end
+end
```
Comment (jlsherrill) [inline app/lib/actions/pulp/repository/ensure_sync_notification.rb]: I actually like having the logic in repo.rb as its much simpler to test there.  I'm more in favor of moving as much logic as we can out of the tasks (aside from the logic which plans the tasks).

Comment (stbenjam) [reply app/lib/actions/pulp/repository/ensure_sync_notification.rb]: Ok, left in the repo.rb.  Can you take a look now?

File: test/controllers/api/v2/repositories_controller_test.rb [hunk 2]
```diff
@@ -40,6 +40,7 @@ def setup
       login_user(User.find(users(:admin).id))
       User.current = User.find(users(:admin).id)
       @request.env['HTTP_ACCEPT'] = 'application/json'
+      ::Katello::Repository.stubs(:ensure_sync_notification)
```
Comment (jlsherrill) [inline test/controllers/api/v2/repositories_controller_test.rb]: i don't think this stub is needed?

File: test/controllers/sync_management_controller_test.rb [hunk 3]
```diff
@@ -17,6 +17,7 @@ def setup
       login_user(User.find(users(:admin).id))
       models
       permissions
+      ::Katello::Repository.stubs(:ensure_sync_notification)
```
Comment (jlsherrill) [inline test/controllers/sync_management_controller_test.rb]: I don't think this is needed anymore either

**Comment by waldenraines** [comment]
[test katello]

**Comment by waldenraines** [comment]
[test katello]

**Comment by waldenraines** [comment]
[test katello]

**Comment by waldenraines** [comment]
[test katello]

**Comment by waldenraines** [comment]
[test katello]

**Comment by stbenjam** [comment]
Once it gets stuck in this kind of test failure, it seems force pushing has more luck in getting it going again

**Comment by stbenjam** [comment]
[test katello]

**Comment by stbenjam** [comment]
[test katello]

**Comment by akofink** [comment]
Squash your commits as a workaround

**Comment by stbenjam** [comment]
That didn't help :-)

**Comment by stbenjam** [comment]
[test katello]

**Comment by stbenjam** [comment]
[test katello]

**Comment by stbenjam** [comment]
[test katello]

**Comment by stbenjam** [comment]
[test katello] :RAGE:

**Comment by stbenjam** [comment]
[test katello] :pray:

**Comment by stbenjam** [comment]
[test katello] :sweet_potato:

**Comment by stbenjam** [comment]
[test katello] :crying_cat_face:

**Comment by stbenjam** [comment]
Hey crying sad cat emoji seemed to convince jenkins to run

**Comment by akofink** [comment]
Idk, maybe we should do this https://github.com/theforeman/foreman-infra/pull/535

**Comment by stbenjam** [comment]
@jlsherrill Green, and removed those 2 stubs

**Pull request merged**